Tbrands 99 product gallery

JackHowa opened the pull request:
## Description

Add product gallery that highlights the first image if toggled on

## Jira Ticket

- [TBRANDS-99](https://arcpublishing.atlassian.net/jira/software/c/projects/TMEDIA/boards/875?modal=detail&selectedIssue=TBRANDS-99&assignee=5e387d6a1b1d910e5dfd7a9b)

## Acceptance Criteria

Product Gallery - Arc Block is available within PageBuilder Editor and can be used to create a PDP.

Product images are available for display from the Arc Commerce Content Source. API data field - schema.productGallery

Product Gallery - Arc Block will have 2 display states which will be configured within PageBuilder Editor:

Featured Image disabled:

This is the default state.

Featured Image enabled:

A user can override the default state (Featured Image disabled) by selecting a radio button that says: Display the product’s featured image.

When Feature Image enabled:

Desktop: a single column large image will populate the first image position of the gallery.

Mobile: the featured image will populate the first position of the gallery.

Product Gallery - Display Logic:

Featured Image disabled:

Maximum of 8 images will display.

Desktop:

Images will display as a 2 column grid.

Even number of images: Renders a 2 column grid

Odd number of images: Renders a 2 column grid. Last image will nest as the last item in the left column. 

Mobile:

Images will appear in a gallery carousel.

Gallery navigation will have a forward button which when clicked will unveil the next image.

Gallery navigation will have a back button which when clicked will unveil the previous image.

Users can horizontally navigate the gallery by swiping from:

right to left which will unveil the next image.

left to right which will unveil the previous image.

On page load:

The forward button is the only button to display.

If a user engages with the forward button or swipes from right to left and unveils the 2nd image, the back button will also display.

If user reaches the last image of the gallery, the forward button will disappear and only the back button will display.

Featured Image enabled:

Maximum of 9 images will display (featured image + up to 8 other product images).

Desktop:

Images will display:

Position #1: a single column large image

Position #2 - #9: will display beneath the single column large image as a 2 column grid.

Even number of images: Renders a 2 column grid

Odd number of images: Renders a 2 column grid. Last image will nest as the last item in the left column. 

Mobile:

Images will appear in a gallery carousel.

The featured image will populate the 1st position of the gallery.

Gallery navigation will have a forward button which when clicked will unveil the next image.

Gallery navigation will have a back button which when clicked will unveil the previous image.

Users can horizontally navigate the gallery by swiping from:

right to left which will unveil the next image.

left to right which will unveil the previous image.

On page load:

The forward button is the only button to display.

If a user engages with the forward button or swipes from right to left and unveils the 2nd image, the back button will also display.

If user reaches the last image of the gallery, the forward button will disappear and only the back button will display.

## Test Steps

### Fusion 

1. Checkout this branch `git checkout TBRANDS-99-product-gallery` for feature pack as well
2. Make sure to link themes components pr as well https://github.com/WPMedia/arc-themes-components/pull/132 -> make sure to reinstall dependencies for themes components for sb
3. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/product-gallery-block,@wpmedia/commerce-product-content-source-block`
4. Create a new page
5. Use the right rail layout
6. Add the product gallery to the main section
7. Toggle the featured image on and off to see the first image displayed at various widths 

desktop
<img width="1437" alt="Screen Shot 2022-08-09 at 15 12 41" src="https://user-images.githubusercontent.com/5950956/183751772-820d078b-3c31-48ff-8c5a-db33a012e2d4.png">
<img width="1440" alt="Screen Shot 2022-08-09 at 15 12 33" src="https://user-images.githubusercontent.com/5950956/183751781-e992aa65-5359-4d12-95eb-25e2bdf4e7bf.png">

mobile

- edge to edge
- I also used the other gallery styles because we weren't following the dots mock

<img width="547" alt="Screen Shot 2022-08-09 at 15 14 53" src="https://user-images.githubusercontent.com/5950956/183752874-407a7338-8fca-4419-a8db-7c144a764fbe.png">
<img width="544" alt="Screen Shot 2022-08-09 at 15 14 58" src="https://user-images.githubusercontent.com/5950956/183752771-b1505590-2153-4649-b028-66d78f160560.png">

### Storybook

1. Look at chromatic at the various breakpoints 
2. You'll notice that the images are being set to be 25vw for non-featured image on desktop.
3. The featured image will be 50vw per the mocks on the screen for desktop
4. The mobile is edge to edge based on the designs 

## Dependencies or Side Effects

- Feature pack pr https://github.com/WPMedia/arc-themes-feature-pack/pull/327
- I updated the mock auth data to make sure that the external url token is being used here in the gallery
- merge themes components pr https://github.com/WPMedia/arc-themes-components/pull/132

## Author Checklist

- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [ ] Confirmed this PR has unit test files
  - [ ] Ran `npm run test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
        please explain why (so that we can fix it whenever it gets refactored).
- [x] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.


# question: 

- What are the dimensions of the designs in the images? will it always be a height and width for the image? or will we use whatever native resolution of the images (may result in one some odd spacing)? -> always 4:3 based on Matt Nash answer
- show dots based on mocks? -> no
- Reuse carousel translation like next slide form product-assortment-carousel into global phrases. Should "products" be a global phrase? is "next" and "previous" ok as global phrases? 
- Can product.item take in a classname? currently I'm selecting the first instance of its class but that item doesn't take in additional classnames
- I used the styling for the other commerce carousels with the icon buttons to make sure that the styling made sense if we were not going to use dots. Is this ok from a design perspective? 

# todos

- [x] Remove the non-modifier from the bem naming of `--disabled` for the top level block. add the base bem name `b-block-name`. Update tests
- [x] Take in id from global data like component context done in other carousel
- [x] Display images within mock data on sb and on the test
- [x] Style the two modifiers for enhanced styles in the theme file for featured image enabled custom setting logic. Make sure the first element styles apply on both breakpoints -- default and desktop. The first element in the list will need a modifier like first-item. (This could be done with a :first or something for css I think but let's make sure it's following other v2 styles precedent if there is one.)  
- [x] Cutoff render of max of 8 items are available to be toggled to or seen with featured disabled, regardless of breakpoint
- [x] Cutoff render of max of 9 items to be toggled or seen with featured enabled, regardless of breakpoint
- [x] Take in label, other translations for carousel from block translations like other blocks
- [x] Localize some like a fallback aria label for Products? 
- [x] Add some basic responsive image widths for resizing with 4:3 aspect ratio
- [x] Investigate passing in a class into the carousel item

JackHowa committed:
feat: Add basic classname check on featured image enabled custom field prop

JackHowa committed:
feat: Handle global content returning not set or an empty object

JackHowa committed:
feat: Update icon based on product feedback

JackHowa committed:
refactor: Rename and extract default custom fields from test

JackHowa committed:
refactor: Rename mock data variable to be more clear that we're assuming some global content will exist here

Carousel div will always render with this ac

JackHowa committed:
feat: Render a carousel component if anything at all

JackHowa committed:
refactor: Remove unnecessary modifier. Use default block class and modify only

JackHowa committed:
feat: Link gallery product import styles for v2

JackHowa committed:
feat: Use component context for id based on other practice in category carousel

JackHowa committed:
ci: Mock get id for component context for potential carousel interaction use

JackHowa committed:
feat: Update to include data as presentational component

JackHowa committed:
fix: Handle carousel items being undefined

Commit:
Merge branch 'arc-themes-release-version-2.0.1' into TBRANDS-99-product-gallery

JackHowa committed:
docs: Add mock content and render two images in the gallery

JackHowa committed:
test: Add carousel test items for rendering list

JackHowa committed:
feat: Add panelling based on desired overflow effect for desktop

JackHowa committed:
style: Fix reorder of the styles

JackHowa committed:
feat: Return only appropriate number of slides per custom field

JackHowa committed:
feat: Add internationalized labels based on previous designs.

Possible improvement could be using global phrases for all but product aria label

JackHowa committed:
feat: Allow image to grow with the srcset and sizes desires

JackHowa committed:
ci: Update themes components

JackHowa committed:
feat: Take in auth token for the external url. Display the resizer sizes for the image at various breakpoints

JackHowa committed:
fix: Update mock data to use external url auth token

JackHowa committed:
fix: Take in resizer url for showing product gallery

Commit:
Merge branch 'arc-themes-release-version-2.0.1' into TBRANDS-99-product-gallery

matthewroach commented on blocks/product-gallery-block/features/product-gallery/default.jsx:
```diff
@@ -0,0 +1,116 @@
+
+const BLOCK_CLASS_NAME = "b-product-gallery";
+
+export function ProductGalleryDisplay(props) {
+	const { arcSite, data, id, isFeaturedImageEnabled, resizerAppVersion, resizerURL } = props;
```
Small code consistency note - we favour unpacking the arguments over destructing them later

```suggestion
export function ProductGalleryDisplay({ arcSite, data, id, isFeaturedImageEnabled, resizerAppVersion, resizerURL }) {
```

JackHowa replied:
ok will do

JackHowa replied:
updated @matthewroach

matthewroach commented on blocks/product-gallery-block/features/product-gallery/default.jsx:
```diff
@@ -0,0 +1,116 @@
+
+	const { locale } = getProperties(arcSite);
+	const phrases = getTranslatedPhrases(locale);
+	const carouselItems = data?.schema?.productGallery?.value?.assets
```
Two items 

1. Should this be done ahead of the presentational component? - That way the current `data` prop doesn't need to know about the nested object location
2. If there is no items we should return `null`

JackHowa replied:
cool will do! thanks

JackHowa replied:
updated @matthewroach

matthewroach commented on blocks/product-gallery-block/features/product-gallery/default.jsx:
```diff
@@ -0,0 +1,116 @@
+							alt={altText}
+							src={url}
+							resizedOptions={{ auth: targetAuth }}
+							width={372}
```
Why choose 372 x 279 as the with and height values?

JackHowa replied:
from the mocks for mobile. would you have a different suggestion? @matthewroach

JackHowa replied:
375 was suggested for the base width and height from product

matthewroach commented on blocks/product-gallery-block/_index.scss:
```diff
@@ -0,0 +1,14 @@
+
+.b-product-gallery {
+	&--featured-image-enabled {
+		.c-carousel__slide:first-child {
```
We shouldn't do this - blocks shouldn't know about components in this way. We should explore another option to achieve what we need here

JackHowa replied:
yeah good note. I was looking into the carousel item taking in a class name. is that a pattern that you'd support? @matthewroach 

in my notes: 

```
Can product.item take in a classname? currently I'm selecting the first instance of its class but that item doesn't take in additional classnames

- [ ] Investigate passing in a class into the carousel item
```

JackHowa replied:
updated. make sure to link the themes components pr https://github.com/WPMedia/arc-themes-components/pull/132

matthewroach reviewed (changes requested):
On desktop the layout isn't as I expect with images not filling the gallery

<img width="1134" alt="Microsoft Edge-2022-08-10-12-55 51" src="https://user-images.githubusercontent.com/868127/183895084-1a20f975-d39b-428c-b4f4-47d23e9a43ff.png">

Commit:
Update blocks/product-gallery-block/features/product-gallery/default.jsx

Co-authored-by: Matthew Roach <matthewroach@users.noreply.github.com>

JackHowa committed:
fix: Featured image should take up full width on desktop regardless of its position in layout

Via https://github.com/WPMedia/arc-themes-blocks/pull/1471#pullrequestreview-1068142490

JackHowa committed:
fix: Handle if no carousel items are shown

Via https://github.com/WPMedia/arc-themes-blocks/pull/1471#discussion_r942357634

JackHowa committed:
Merge branch 'TBRANDS-99-product-gallery' of github.com:WPMedia/fusion-news-theme-blocks into TBRANDS-99-product-gallery

JackHowa committed:
Merge branch 'arc-themes-release-version-2.0.1' into TBRANDS-99-product-gallery

JackHowa committed:
fix: Use new aspect ratio 1:1 based on new requirements. Fix featured image filling full screen based on width undefined logic

JackHowa committed:
fix: Update responsive images array for 375 change

JackHowa committed:
docs: Update mock carousel items in storybook based on presentational view

JackHowa commented:
> On desktop the layout isn't as I expect with images not filling the gallery

Updated to be full width

JackHowa committed:
refactor: Pass in class for carousel item component update

via https://github.com/WPMedia/arc-themes-blocks/pull/1471#discussion_r942359422

JackHowa committed:
test: Remove relevant enabled featured slide pending change to components

JackHowa committed:
docs: Add more items to the carousel items storybook

JackHowa committed:
feat: Add more items to the gallery array list for testing

JackHowa committed:
refactor: Make sure that the presentational component does the slicing for the sake of storybook and carousel items length check

JackHowa committed:
ci: Update theme components

JackHowa commented:
@matthewroach updated so you don't need to link the theme components anymore

matthewroach commented on blocks/product-gallery-block/index.story.jsx:
```diff
@@ -0,0 +1,122 @@
+	title: "Blocks/Product Gallery",
+	parameters: {
+		chromatic: { viewports: [320, 1200] },
+	},
```
We should add the following here to ensure chromatic has the correct output

```
		cssVariables: {
			theme: "commerce",
		},
```

JackHowa replied:
will do, thanks. just noticed you did this before

JackHowa replied:
updated -- thanks

JackHowa committed:
ci: Update to use css variable on theme commerce

JackHowa committed:
ci: Fix use parameters for variables

JackHowa commented:
Blocked based on design and ac discussion around back and forward buttons on mobile

JackHowa committed:
feat: Remove next and previous button arrangements.

Based on discussion with Matt and Kelly

JackHowa commented:
@haganka updated to remove the next and previous accessibility buttons for mobile. thanks for the quick feedback!

JackHowa committed:
Merge branch 'arc-themes-release-version-2.0.1' into TBRANDS-99-product-gallery

JackHowa committed:
fix: Update intermediary variable fallback
